[SPARK-11212][Core][Streaming]Make preferred locations support ExecutorCacheTaskLocation and update…

zsxwing opened the pull request:
… ReceiverTracker and ReceiverSchedulingPolicy to use it

This PR includes the following changes:
1. Add a new preferred location format, `executor_<host>_<executorID>` (e.g., "executor_localhost_2"), to support specifying the executor locations for RDD.
2. Use the new preferred location format in `ReceiverTracker` to optimize the starting time of Receivers when there are multiple executors in a host. 

The goal of this PR is to enable the streaming scheduler to place receivers (which run as tasks) in specific executors. Basically, I want to have more control on the placement of the receivers such that they are evenly distributed among the executors. We tried to do this without changing the core scheduling logic. But it does not allow specifying particular executor as preferred location, only at the host level. So if there are two executors in the same host, and I want two receivers to run on them (one on each executor), I cannot specify that. Current code only specifies the host as preference, which may end up launching both receivers on the same executor. We try to work around it but restarting a receiver when it does not launch in the desired executor and hope that next time it will be started in the right one. But that cause lots of restarts, and delays in correctly launching the receiver.

So this change, would allow the streaming scheduler to specify the exact executor as the preferred location. Also this is not exposed to the user, only the streaming scheduler uses this.

zsxwing committed:
Make preferred locations support ExecutorCacheTaskLocation and update ReceiverTracker and ReceiverSchedulingPolicy to use it

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

zsxwing commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverSchedulingPolicy.scala:
```diff
@@ -69,9 +69,12 @@ private[streaming] class ReceiverSchedulingPolicy {
   def scheduleReceivers(
-      receivers: Seq[Receiver[_]], executors: Seq[String]): Map[Int, Seq[String]] = {
+      receivers: Seq[Receiver[_]],
+      executors: Seq[ExecutorCacheTaskLocation]): Map[Int, Seq[TaskLocation]] = {
```
Use `TaskLocation` in the return type because the locations could be host from `Receiver.preferredLocation`, or `ExecutorCacheTaskLocation`.

SparkQA commented:
**[Test build #43982 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43982/consoleFull)** for PR 9181 at commit [`f5e7c4f`](https://github.com/apache/spark/commit/f5e7c4f0fd24efd179b96725f8b5b0ed3450b4b2).

zsxwing commented:
I tested this patch using 5 workers, 24 executors, 24 receivers and there were no receiver restarting logs in the test.

SparkQA commented:
**[Test build #43982 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43982/consoleFull)** for PR 9181 at commit [`f5e7c4f`](https://github.com/apache/spark/commit/f5e7c4f0fd24efd179b96725f8b5b0ed3450b4b2).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43982/
Test FAILed.

tdas commented on core/src/main/scala/org/apache/spark/scheduler/TaskLocation.scala:
```diff
@@ -65,7 +70,15 @@ private[spark] object TaskLocation {
+      } else {
+        new HostTaskLocation(str)
+      }
     } else {
```
@andrewor14 @kayousterhout 
Can you take a look at this change?

kayousterhout replied:
Is this necessary because previously, we didn't allow the user to pass in ExecutorCacheTaskLocations and we just tried to figure them out automatically? (and why doesn't that work here?)

tdas replied:
The goal of this PR is to enable the streaming scheduler to place receivers (which run as tasks) in specific executors. Basically, I want to have more control on the placement of the receivers such that they are evenly distributed among the executors. We tried to do this without changing the core scheduling logic, but its not possible to precisely put a task in a specific executor with this change. I dont think this change affects the performance of anything else. 

Also this is not exposed to the user. Only the streaming scheduler uses this.

kayousterhout replied:
Ok sounds good -- can you add this description to the JIRA and to the pull request description (so that it will be in the commit message)?  Scheduler changes LGTM.

tdas replied:
@zsxwing Please do so. :)

zsxwing commented:
retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #44022 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44022/consoleFull)** for PR 9181 at commit [`f5e7c4f`](https://github.com/apache/spark/commit/f5e7c4f0fd24efd179b96725f8b5b0ed3450b4b2).

SparkQA commented:
**[Test build #44022 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44022/consoleFull)** for PR 9181 at commit [`f5e7c4f`](https://github.com/apache/spark/commit/f5e7c4f0fd24efd179b96725f8b5b0ed3450b4b2).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44022/
Test FAILed.

zsxwing commented:
retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #44028 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44028/consoleFull)** for PR 9181 at commit [`f5e7c4f`](https://github.com/apache/spark/commit/f5e7c4f0fd24efd179b96725f8b5b0ed3450b4b2).

SparkQA commented:
**[Test build #44028 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44028/consoleFull)** for PR 9181 at commit [`f5e7c4f`](https://github.com/apache/spark/commit/f5e7c4f0fd24efd179b96725f8b5b0ed3450b4b2).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44028/
Test FAILed.

zsxwing commented:
retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #44035 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44035/consoleFull)** for PR 9181 at commit [`f5e7c4f`](https://github.com/apache/spark/commit/f5e7c4f0fd24efd179b96725f8b5b0ed3450b4b2).

SparkQA commented:
**[Test build #44035 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44035/consoleFull)** for PR 9181 at commit [`f5e7c4f`](https://github.com/apache/spark/commit/f5e7c4f0fd24efd179b96725f8b5b0ed3450b4b2).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44035/
Test FAILed.

srowen commented:
@zsxwing I think this duplicated the existing issue / PR https://github.com/apache/spark/pull/9096 . Does this cover the same logic, and, can you include all of the changes here?

zsxwing commented:
retest this please

zsxwing commented:
> @zsxwing I think this duplicated the existing issue / PR #9096 . Does this cover the same logic, and, can you include all of the changes here?

This is a different issue that only about `ExecutorCacheTaskLocation`.  #9096 is a minor patch. I can update this one after #9096 gets merged.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #44048 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44048/consoleFull)** for PR 9181 at commit [`f5e7c4f`](https://github.com/apache/spark/commit/f5e7c4f0fd24efd179b96725f8b5b0ed3450b4b2).

SparkQA commented:
**[Test build #44048 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44048/consoleFull)** for PR 9181 at commit [`f5e7c4f`](https://github.com/apache/spark/commit/f5e7c4f0fd24efd179b96725f8b5b0ed3450b4b2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44048/
Test PASSed.

srowen commented:
@zsxwing I merged #9096

zsxwing committed:
Merge branch 'master' into executor-location

zsxwing committed:
Add a unit test

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #44148 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44148/consoleFull)** for PR 9181 at commit [`35f7936`](https://github.com/apache/spark/commit/35f7936ac38551d13c0dc253c05d63a95a4ff30d).

SparkQA commented:
**[Test build #44148 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44148/consoleFull)** for PR 9181 at commit [`35f7936`](https://github.com/apache/spark/commit/35f7936ac38551d13c0dc253c05d63a95a4ff30d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44148/
Test PASSed.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverSchedulingPolicy.scala:
```diff
@@ -30,22 +30,22 @@ import org.apache.spark.util.Utils
  * - The first phase is global scheduling when ReceiverTracker is starting and we need to schedule
  *   all receivers at the same time. ReceiverTracker will call `scheduleReceivers` at this phase.
  *   It will try to schedule receivers with evenly distributed. ReceiverTracker should update its
```
nit: grammar. It will try to scheduler receiver such that they are evenly distributed

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverSchedulingPolicy.scala:
```diff
@@ -30,22 +30,22 @@ import org.apache.spark.util.Utils
- *   receiverTrackingInfoMap according to the results of `scheduleReceivers`.
- *   `ReceiverTrackingInfo.scheduledExecutors` for each receiver will set to an executor list that
+ *   `receiverTrackingInfoMap` according to the results of `scheduleReceivers`.
+ *   `ReceiverTrackingInfo.scheduledLocations` for each receiver will set to an location list that
```
nit: will set --> should be set

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverSchedulingPolicy.scala:
```diff
@@ -163,40 +169,46 @@ private[streaming] class ReceiverSchedulingPolicy {
+    // handle this case
+    scheduledLocations ++= preferredLocation.map(TaskLocation(_))
+
+    val executorWeights: Map[ExecutorCacheTaskLocation, Double] =
```
nit: would be good to put a `{...}` around this whole thing.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverSchedulingPolicy.scala:
```diff
@@ -163,40 +169,46 @@ private[streaming] class ReceiverSchedulingPolicy {
+
+    val executorWeights: Map[ExecutorCacheTaskLocation, Double] =
+      receiverTrackingInfoMap.values.flatMap { receiverTrackingInfo =>
+        receiverTrackingInfo.state match {
```
can you make this a separation method, this is getting pretty big and hard to read.
and add comments.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTrackingInfo.scala:
```diff
@@ -47,7 +48,7 @@ private[streaming] case class ReceiverTrackingInfo(
     name.getOrElse(""),
     state == ReceiverState.ACTIVE,
-    location = runningExecutor.getOrElse(""),
+    location = runningExecutor.map(_.host).getOrElse(""),
```
Its probably a good idea to also add executor information, not just the host.

tdas commented on streaming/src/test/scala/org/apache/spark/streaming/scheduler/ReceiverSchedulingPolicySuite.scala:
```diff
@@ -20,73 +20,102 @@ package org.apache.spark.streaming.scheduler
+      ExecutorCacheTaskLocation("host3", "3"),
+      ExecutorCacheTaskLocation("host4", "4"),
+      ExecutorCacheTaskLocation("host5", "5")
+    )
```
I this can be replaced with `(1 to 5).map { i => ExecutorCacheTaskLocation(s"host$i", s"$i") }`

tdas commented:
Overall, LGTM, except a few minor refactorings.

zsxwing committed:
Minor refactorings to address comments

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

zsxwing commented:
@tdas addressed your comments.

> Its probably a good idea to also add executor information, not just the host.

For this one, I prefer to add the executor info in a separate PR and also add it to UI.

SparkQA commented:
**[Test build #44390 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44390/consoleFull)** for PR 9181 at commit [`63c66eb`](https://github.com/apache/spark/commit/63c66eb231b034820ace712299946ebaef7c922e).

zsxwing commented:
> For this one, I prefer to add the executor info in a separate PR and also add it to UI.

Created https://issues.apache.org/jira/browse/SPARK-11333 to track it

SparkQA commented:
**[Test build #44390 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44390/consoleFull)** for PR 9181 at commit [`63c66eb`](https://github.com/apache/spark/commit/63c66eb231b034820ace712299946ebaef7c922e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44390/
Test PASSed.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverSchedulingPolicy.scala:
```diff
@@ -163,40 +169,59 @@ private[streaming] class ReceiverSchedulingPolicy {
+        // The probability that a scheduled receiver will run in an executor is
+        // 1.0 / scheduledLocations.size
+        scheduledLocations.filter(_.isInstanceOf[ExecutorCacheTaskLocation]).
+          map { location =>
```
nit:shouldnt this map be in the line above? I think `.map { loc =>` would fit.

zsxwing committed:
Fix the style

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #44420 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44420/consoleFull)** for PR 9181 at commit [`46c5197`](https://github.com/apache/spark/commit/46c5197c946cbed75ac28e1b60e64266ca070ee1).

SparkQA commented:
**[Test build #44420 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44420/consoleFull)** for PR 9181 at commit [`46c5197`](https://github.com/apache/spark/commit/46c5197c946cbed75ac28e1b60e64266ca070ee1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44420/
Test PASSed.

tdas commented:
Merging this to master. Thanks @zsxwing
